src/main/java/org/glite/authz/pep/pip/provider/policynamespip/UpdatingPolicyNamesCache.java

```diff
@@ -0,0 +1,210 @@
+     * @see #UPDATEINTERVAL
+     */
+    public void setUpdateInterval(long msecs)    {
+	update_interval=msecs;
+    }
+    
+    ////////////////////////////////////////////////////////////////////////
+    // Constructors
+    ////////////////////////////////////////////////////////////////////////
+    
+    /**
+     * constructs new UpdatingPolicyNamesCache based on given trustDir and
+     * updateInterval
+     * @param trustDir directory containing info files
+     * @throws IOException on read errors for trust_dir or one of the info files
+     * @see #UpdatingPolicyNamesCache(String)
+     * @throws IOException in case of I/O errors
+     */
+    public UpdatingPolicyNamesCache(String trustDir, long updateInterval) throws IOException {
+	this(trustDir);
+	this.update_interval = updateInterval;
+    }
+
+    /**
+     * constructs new UpdatingPolicyNamesCache based on given trustDir and
+     * default {@link #UPDATEINTERVAL}.
+     * @param trustDir directory containing info files
+     * @param updateInterval interval (msec) between info file cache updates
+     * @see #UpdatingPolicyNamesCache(String, long)
+     * @throws IOException in case of I/O errors
+     */
+    public UpdatingPolicyNamesCache(String trustDir) throws IOException {
+	this.cache = new PolicyNamesCache(trustDir);
+    }
+
+
+    ////////////////////////////////////////////////////////////////////////
+    // Main instance method
+    ////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Tries to find given subjectDN in the info files in trust dir.
+     * @param dn String subject DN to look for
+     * @return array of String with all the matching info files
+     * @throws IOException upon reading errors in updating the
+     * {@link PolicyNamesCache}
+     * @see PolicyNamesCache#matchIssuerDN(String)
+     */
+    public String[] findIssuerDN(String dn) throws IOException   {
+	// Update the cache (when needed)
+	updateCache();
+
+	// Obtain read lock
+	read.lock();
+	try {
+	    // Protect against empty cache
+	    if (cache == null)  {
+		log.warn("Encountered empty cache while matching DN "+dn);
+		return new String[0];
+	    }
+
+	    // Find and return matching info files
+	    return cache.matchIssuerDN(dn);
+	} finally {
+	    // release read lock
+	    read.unlock();
+	}
+    }
+    
+
+    ////////////////////////////////////////////////////////////////////////
+    // Private methods
+    ////////////////////////////////////////////////////////////////////////
+
+    /**
+     * Update the internal {@link PolicyNamesCache} when needed
+     * @throws IOException upon I/O errors in updating the
+     * {@link PolicyNamesCache}
+     */
+    private void updateCache() throws IOException    {
+	PolicyNamesCache newCache=null;
+
+	// First check (with appropriate read lock) whether we need to do
+	// anything. If so, create a new cache.
+	read.lock();
+	try {
```

Comment:
I would drop this in favour of a separate thread that just takes the write lock according to a specified period